### PR TITLE
Background music

### PR DESCRIPTION
**Pull request by ShahafPinto**
Hi :)

I added a playlist according to the settings in the issue #53 

Except the next song logic

### CODE REVIEW

**Comment by eternaleclipse** [comment]
Hey, thanks for the PR! 🙌

There are several things that need to be addressed before we can merge the changes.
1. Please run `prettier` on the code for formatting.
2. Please remove the `console.log()` (or make it more descriptive 🙂)
3. I think it would be better if the playlist constants could be declared in an external (ex. `musicAssets.ts` file) and imported from there.
4. There should be a new "Music On/Off" button that is separate from the "Sound" button.
5. As per the project's MVVM architecture, any UI element (buttons, the user sees and interacts with directly should be implemented in the corresponding `View` component (in `src/views/App/...`). The `ViewModel` holds the state and tells the `View` when to update - which in turn, does the actual drawing, plays sound, music etc. An example for this can be found in `src/views/App/Board/Board.tsx`, where sound is played on each piece move. A similar solution can be used for the background music.
6. Currently, the music only starts when `PreferencesViewModel.toggleSoundMode()` is called (So, when the game starts, even when the sound is on, the music doesn't play). This will be solved when the previous point is addressed, and `audio.play()` is moved to the View.
7. I see you have laid some groundwork for the playlist feature in `playNext` - please comment the parts that are not required for the currently functioning code.
8. In `playNext`, there is a code block that is duplicated and can be taken out of the `if` statement:
```js
if (this.soundMode) {
      if (this.audio) {
        this.audio.pause();
      }
      ...
    } else {
      if (this.audio) {
        this.audio.pause();
      }
    }
  };
```
9. In `playNext`, if there is an error in the audio promise, it is not handled - you can add a `console.log()` describing the error, and return from the function.
10. In the `playNext` function, when the end of the playlist is reached, the function resets the counter and calls itself recursively. This can be resolved by re-organizing the function's logic, so that it determines which track to play and sets the counter before proceeding.

**Comment by ShahafPinto** [comment]
Hi, all done :)
I'd be happy if you could go over it and merge if possible.

**Comment by ShahafPinto** [comment]
Hi :) all done, except the part of the auto-play music when rendering, because the Autoplay policy in Chrome.

File: src/views/App/BgMusic/BgMusic.tsx
Comment (yaelih) [inline src/views/App/BgMusic/BgMusic.tsx:36]: This won't work. It should be a function call (missed the parentheses), or even better to remove the anonymous function and just write playcurrent:
instead of () => { playcurrent() } just put playcurrent,

File: src/views/App/BgMusic/BgMusic.tsx
Comment (yaelih) [inline src/views/App/BgMusic/BgMusic.tsx:35]: when having an addEventListener, it's better to have also a removeEventListener somewhere in the code.

File: src/views/App/BgMusic/BgMusic.tsx
Comment (yaelih) [inline src/views/App/BgMusic/BgMusic.tsx:41]: this line of code can be removed. It doesn't seem to do anything.

**Review by yaelih — CHANGES_REQUESTED** [review]
I'm giving a first review here. Wait for maintainers approval before changing anything.

In addition to the feedback directly in code, I'm raising also 2 issues here:

1. When having soundMode off and clicking play/pause, nothing will happen and it can be a bit strange to the user. Maybe an improvement will be to disable and/or have an explanation tooltip, because it's not clear it depends on SoundMode state.

2. Having only an initial value of isPlay in preferencesViewModel, but manipulating it in BgMusic component, seems a bit inaccurate to me.  I assumed we shall have a toggle function in preferencesViewModel at least, otherwise there's no need of declaring it in preferencesViewModel. Am I right?

File: src/services/preferencesViewModel.ts
Comment (ShahafPinto) [reply src/services/preferencesViewModel.ts:25]: Hi, thanks for the comment, i updated

File: src/services/preferencesViewModel.ts
Comment (ShahafPinto) [reply src/services/preferencesViewModel.ts:31]: Hi, thanks for the comment, i updated

File: src/views/App/BgMusic/BgMusic.tsx
Comment (ShahafPinto) [reply src/views/App/BgMusic/BgMusic.tsx:36]: Hi, thanks for the comment, i updated

File: src/views/App/BgMusic/BgMusic.tsx
Comment (ShahafPinto) [reply src/views/App/BgMusic/BgMusic.tsx:35]: Hi, thanks for the comment, i updated

File: src/views/App/BgMusic/BgMusic.tsx
Comment (ShahafPinto) [reply src/views/App/BgMusic/BgMusic.tsx:41]: Hi, thanks for the comment, i updated

**Comment by ShahafPinto** [comment]
Hi, i updated everything except this comment:

2. Having only an initial value of isPlay in preferencesViewModel, but manipulating it in BgMusic component, seems a bit 
 inaccurate to me. I assumed we shall have a toggle function in preferencesViewModel at least, otherwise there's no need of 
 declaring it in preferencesViewModel. Am I right?

I thout it was right to put this value in preferencesViewModel because it's the status of one of the buttons (in the view).
could you please tell me what you think is the correct thing to do?

**Comment by ShahafPinto** [comment]
Hi, 

1. Music Play/Pause should not be dependant upon SoundMode On/Off  - done
2. Music button not in the same line visually - done
3. Music button should be part of ActionButtons component and not in it's seperate component - done
4. isMusicPlaying should be updated in preferencesViewModel.ts (during button click) - done
5. During git commits, clean and remove changes that are not important (e.g. extra blanks lines etc) (?)	 - done
6. formatting (or just use prettier) - done
7. playcurrent => playCurrentTrack (camelCase, better naming) - done
8. isPlay - better naming (probably better to just import the original variable - isMusicPlaying) - done
9. add semicolons (prettier fixes this) - done
10. playPauseClick => playPauseClicked / handler - done
11. currentSong is always reset to -1  - didn't do that because I didn't insert a Pause-button
12. addEventListener / removeElementListener currently receive an anonymous wrapper around playcurrent. During removeEventListener, it needs to receive a reference to the *same* function used during addEventListener. So, we need to use the actual function and not a wrapper (in both function). - done
13. The current setup will not work if we reuse or re-render the component. The reason being is that the function being passed to addEventListener and removeEventListener are recreated every render. To solve this, we can use `useCallback` to "cache" the function. - done
14.  In playPauseClick, there is this line `isPlay = !isPlay`.
```
const playPauseClick = ()=>{
        currentSong = -1;
        if (isPlay && audio) {
            audio.pause();
            audio.removeEventListener("ended", ()=>playcurrent())
            isPlay = !isPlay 
        } 
        else if(soundMode){
            playcurrent();
            isPlay = !isPlay
        }
    }
```

If we keep this line inside the `if` statements, we could use constant values like `isPlay = false` for clarity. Alternatively, we can move this line outside the if clasues so it switches to the opposite value in any case.
- done
15. Instead of creating a new Audio element, we can use `audio.src = "http:/..."`. This also eliminates the need for calling audio.pause() in NextClick, and adding a new eventListener every time. In this setup, adding the listener will happen only during initialization, and removing the listener will happen only once - during component destruction. A better way to solve this is using `useEffect` with an empty dependency array, and a return function that removes the listener. We can work it out together :) - done
16. getCurr => getNextSongNum - done
17. In preferencesViewModel, isMusicPlaying should use localStorage (we need to discuss this because we may have found there is support for autoplay without user interaction - TBD) - done
18. In MVVM, we separate the ViewModel (logic representation of the View) and the View (UI) itself. This is done to enable better testing. So, anything involving state (such as isPlay and currentSong) should not be in View but in the parallel ViewModel class. In this case, we can move currentSong to preferencesViewModel, since it's not too large - in cases of larger state management, it's better to create it's own class. Note that the MVVM architecture is not common in React (and used in this project mainly for experimental and learning reasons). - not sure if it's done

Topics to research later:
1. audio plays without interaction in JSFiddle
2. Is `isPlay` a reference to preferenceViewModel.isMusicPlaying? Does it update the original variable? - done

**Comment by eternaleclipse** [comment]
The fix to `useEffect` is perfect!

There is still one problem related to comment 11 from [here](https://github.com/chess-mazes/chess-mazes/pull/80#issuecomment-1587425595) - In Chrome, when starting the music and refreshing the page, the music button shows "playing" but since the user hasn't interacted with on anything yet, the console shows the error

**Comment by eternaleclipse** [comment]
I've added a small set of changes to your code.
- `musicMode` is now directly taken from `preferencesViewModel`
- The exception handler now checks the error type to see if it is to turn off the `musicMode` correctly
- The exception handler throws the error if it doesn't match `NotAllowedError` or `SecurityError`
- `musicMode` is correctly being updated and reflected back to the view

It works beautifully - merging.

@ShahafPinto thank you so much for your work on this PR. Your persistence and hard work is exceptional!
@yaelih Thank you for your tremendous help with this PR!

Cheers 🍻

**Pull request merged**
